### PR TITLE
Reduce code duplication between String.match and search

### PR DESCRIPTION
**Pull request by mmatyas**
Moved the common preparation code for 'search' and 'match'
to a separate function.

JerryScript-DCO-1.0-Signed-off-by: Mátyás Mustoha mmatyas@inf.u-szeged.hu

### CODE REVIEW

**Comment by zherczeg** [comment]
How much is the gain?

**Comment by mmatyas** [comment]
About 35 lines.

**Comment by zherczeg** [comment]
Good reply :D 

I mean binary size reduction. Since this is an optimization patch I am curious about its effect.

**Comment by mmatyas** [comment]
Ah ok, here's what I see using the default build parameters:

before:

File | Size
--:|:--
build/bin/jerry | 217880
build/lib/libjerry-core.a | 3301510
build/lib/libjerry-ext.a | 141244
build/lib/libjerry-libc.a | 63484
build/lib/libjerry-libm.a | 139790
build/lib/libjerry-port-default.a | 32248
build/lib/libjerry-port-default-minimal.a | 28948

after:

File | Size
--:|:--
build/bin/jerry | 217880
build/lib/libjerry-core.a | 3302310
build/lib/libjerry-ext.a | 141244
build/lib/libjerry-libc.a | 63468
build/lib/libjerry-libm.a | 139758
build/lib/libjerry-port-default.a | 32264
build/lib/libjerry-port-default-minimal.a | 28956

Seems there's no real size gain :(

**Comment by zherczeg** [comment]
No problem. It is still a good patch.

**Comment by mmatyas** [comment]
Rebased to master to fix the Travis build.

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c [hunk 1]
```diff
@@ -406,26 +406,52 @@ ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this arg
...
+ */
+static ecma_value_t
+ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this argument */
+                                            ecma_value_t regexp_arg) /**< routine's argument */
+{
+  ecma_value_t this_to_string_value = ECMA_VALUE_EMPTY;
+  ecma_value_t regexp_value = ECMA_VALUE_EMPTY;
+
+  ecma_value_t ret_value = ecma_builtin_string_prepare_search (this_arg, &this_to_string_value,
+                                                               regexp_arg, &regexp_value);
```
Comment (LaszloLango) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: I think we can reduce the number of parameters of `ecma_builtin_string_prepare_search`. `this_to_string_value` can be returned by the function if no error happened.

Comment (mmatyas) [reply jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: How should we handle the errors produced in `ecma_builtin_string_prepare_search`? Is there a way to decide whether the returned `this_to_string_value` contains an error or a valid value?

Comment (LaszloLango) [reply jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: Yes, you can use `ECMA_IS_VALUE_ERROR`

Comment (mmatyas) [reply jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: Well I've started looking into this, but honestly this makes the code ugly and less readable to me. Personally I'd still prefer the current 2 input + 2 output + 1 retval setup.

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c [hunk 2]
```diff
@@ -382,17 +382,17 @@ ecma_builtin_string_prototype_object_locale_compare (ecma_value_t this_arg, /**<
...
+ * of the String prototype.
  *
  * @return ecma value
  *         Returned value must be freed with ecma_free_value.
  */
 static ecma_value_t
-ecma_builtin_string_prototype_object_match (ecma_value_t this_arg, /**< this argument */
-                                            ecma_value_t arg) /**< routine's argument */
+ecma_builtin_string_prepare_search (ecma_value_t this_arg, /**< this argument */
+                                    ecma_value_t *this_to_string_value_ptr, /**< ptr to store as string */
```
Comment (LaszloLango) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: `/**< [out] pointer to store as string */`

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c [hunk 3]
```diff
@@ -382,17 +382,17 @@ ecma_builtin_string_prototype_object_locale_compare (ecma_value_t this_arg, /**<
...
+                                    ecma_value_t regexp_arg, /**< regex argument */
+                                    ecma_value_t *regexp_value) /**< ptr to store the regexp object */
```
Comment (LaszloLango) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: /**< [out] pointer to store the regexp object */

File: jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c [hunk 4]
```diff
@@ -382,17 +382,17 @@ ecma_builtin_string_prototype_object_locale_compare (ecma_value_t this_arg, /**<
...
 
 /**
- * The String.prototype object's 'match' routine
- *
- * See also:
- *          ECMA-262 v5, 15.5.4.10
+ * The common preparation code for 'search' and 'match' functions
+ * of the String prototype.
  *
  * @return ecma value
```
Comment (LaszloLango) [inline jerry-core/ecma/builtin-objects/ecma-builtin-string-prototype.c]: Add more description to the return value

**Pull request merged**